### ISSUE TITLE
Setting highlight query on Inner Hits throwing NullPointerException

### ISSUE THREAD

**Issue by AliNield** [elastic/elasticsearch#21061]
**Elasticsearch version**: 2.4.1

**Description of the problem including expected versus actual behavior**:
When setting a Highlight query on an Inner Hits query, it throws a NullPointerException. We want to be able to set a different highlight query for the Inner hits than the main query. 

This is a bit of a forced example, but it is similar to what we need it to do. The documentation describes that Inner Hits support highlighting, so I would expect it to support setting the highlight query. https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-inner-hits.html

**Steps to reproduce**:
 1. Add a parent/child mapping
```
curl -XPOST localhost:9200/test_index -d '{
    "settings" : {
        "number_of_shards" : 1
    },
    "mappings": {
    "profile": {
      "properties": {
          "id": {
          "type": "string",
        },
        "name": {
          "type": "string"
        },        
        "summary": {
          "type": "string"
        }
      }
    },
    "tweet": {
      "_parent": {
        "type": "profile"
      },
      "properties": {
        "id": {
          "type": "string",
        },        
        "body": {
          "type": "string"
        }
      }
    }
  }
}'

```
 2. Add some data
```
curl -XPUT localhost:9200/test_index/profile/1 -d '{
  "name" : "Bob",
  "summary" : "The quick brown fox"
}'


curl -XPUT localhost:9200/test_index/tweet/100?parent=1 -d '{
  "body" : "I like lego"
}'

curl -XPUT localhost:9200/test_index/tweet/200?parent=1 -d '{
  "body" : "going to build some lego"
}'
```
 3. Make the search
```
curl -XPOST localhost:9200/test_index/_search -d '{
   "query":{
      "has_child":{
         "type":"tweet",
         "query":{
            "match":{
              "body" : "build"
            }
         },
         "inner_hits":{
            "highlight":{
               "fields":{
                  "body":{
                     "highlight_query":{
                        "match":{
                           "body":"lego"
                        }
                     }
                  }
               }
            }
         }
      }
   }
}'
```

**Provide logs (if relevant)**:

```
{
  "error": {
    "root_cause": [
      {
        "type": "null_pointer_exception",
        "reason": null
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query_fetch",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "test_index",
        "node": "kQJBHiijQCW_3FsdFssnrQ",
        "reason": {
          "type": "null_pointer_exception",
          "reason": null
        }
      }
    ]
  },
  "status": 500
}
```

Thanks

**Comment by cbuescher** [comment]
I can reproduce this on 2.4.1 but it seems to be fixed in 5.0. The NPE I see in the log with the above example on 2.4.1 is:
```
[...]
	at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:386)
	at org.elasticsearch.action.search.SearchPhaseExecutionException.guessRootCauses(SearchPhaseExecutionException.java:152)
	at org.elasticsearch.action.search.SearchPhaseExecutionException.getCause(SearchPhaseExecutionException.java:99)
	at java.lang.Throwable.printStackTrace(Throwable.java:665)
	at java.lang.Throwable.printStackTrace(Throwable.java:721)
	at org.apache.log4j.DefaultThrowableRenderer.render(DefaultThrowableRenderer.java:60)
	at org.apache.log4j.spi.ThrowableInformation.getThrowableStrRep(ThrowableInformation.java:87)
	at org.apache.log4j.spi.LoggingEvent.getThrowableStrRep(LoggingEvent.java:413)
	at org.apache.log4j.WriterAppender.subAppend(WriterAppender.java:313)
	at org.apache.log4j.WriterAppender.append(WriterAppender.java:162)
	at org.apache.log4j.AppenderSkeleton.doAppend(AppenderSkeleton.java:251)
	at org.apache.log4j.helpers.AppenderAttachableImpl.appendLoopOnAppenders(AppenderAttachableImpl.java:66)
	at org.apache.log4j.Category.callAppenders(Category.java:206)
	at org.apache.log4j.Category.forcedLog(Category.java:391)
	at org.apache.log4j.Category.log(Category.java:856)
	at org.elasticsearch.common.logging.log4j.Log4jESLogger.internalWarn(Log4jESLogger.java:135)
	at org.elasticsearch.common.logging.support.AbstractESLogger.warn(AbstractESLogger.java:109)
	at org.elasticsearch.rest.BytesRestResponse.convert(BytesRestResponse.java:134)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:96)
	at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:87)
	at org.elasticsearch.rest.action.support.RestActionListener.onFailure(RestActionListener.java:60)
	at org.elasticsearch.action.support.TransportAction$1.onFailure(TransportAction.java:95)
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.raiseEarlyFailure(AbstractSearchAsyncAction.java:294)
	... 10 more
Caused by: java.lang.NullPointerException
	at org.elasticsearch.index.query.QueryParseContext.parseInnerQuery(QueryParseContext.java:251)
	at org.elasticsearch.index.query.IndexQueryParserService.innerParse(IndexQueryParserService.java:320)
	at org.elasticsearch.index.query.IndexQueryParserService.parse(IndexQueryParserService.java:223)
	at org.elasticsearch.index.query.IndexQueryParserService.parse(IndexQueryParserService.java:218)
	at org.elasticsearch.search.query.QueryParseElement.parse(QueryParseElement.java:33)
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:856)
	at org.elasticsearch.search.SearchService.createContext(SearchService.java:667)
	at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:633)
	at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:472)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:392)
	at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:389)
	at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:77)
	at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
	... 3 more
```

**Comment by cbuescher** [comment]
This is fixed now on the 2.4 branch with #21065. On 5.0 the problem doesn't appear any more due to a large change in the query parsing infrastructure.
